### PR TITLE
fix for ubuntu kernel 6.6+

### LINKED ISSUE #780: No such file or directory - unprivileged_userns_clone
Am seeing the following error when attempting to install Sysbox on an RKE2 worker node.

`sysctl: cannot stat /proc/sys/kernel/unprivileged_userns_clone: No such file or directory`

---
Using the latest version of all system components, including the latest "longterm" version of the Linux kernel as per https://kernel.org.

`os=Ubuntu 22.04.4 LTS, kernel=6.6.17-060617-generic, sysbox=0.6.3 CE, rke2=v1.28.6+rke2r1, rancher=v2.8.2`

---
Here is a summary of results when using various kernel versions.

- 6.8-rc4 (mainline) - fails
- 6.7.5 (stable) - fails
- 6.6.17 (longterm) - fails
- 6.5.18 (from the mainline) - succeeds
- 6.5.18 (Ubuntu LTS HWE) - succeeds

---

Here are the Sysbox logs.

> Detected **Kubernetes version v1.28**
2024-02-18T14:06:20.436036364+10:00 Adding K8s taint "sysbox-runtime=not-running:NoSchedule" to node ...
2024-02-18T14:06:20.502085520+10:00 node/w6 modified
2024-02-18T14:06:20.506327963+10:00 Adding K8s label "crio-runtime=installing" to node ...
node/w6 not labeled
2024-02-18T14:06:20.571532866+10:00 Deploying CRI-O installer agent on the host (v1.28) ...
2024-02-18T14:06:21.002092200+10:00 Running CRI-O installer agent on the host (may take several seconds) ...
2024-02-18T14:06:23.395783002+10:00 Removing CRI-O installer agent from the host ...
2024-02-18T14:06:23.908565022+10:00 Configuring CRI-O ...
2024-02-18T14:06:25.114476735+10:00 Adding K8s label "sysbox-runtime=installing" to node ...
2024-02-18T14:06:25.175698518+10:00 node/w6 not labeled
2024-02-18T14:06:25.194719353+10:00 Installing Sysbox dependencies on host ...
2024-02-18T14:06:25.205312310+10:00 Copying shiftfs sources to host ...
2024-02-18T14:06:25.206274620+10:00 **Kernel version 6.6**, which is above the max required for shiftfs (6.2); skipping shiftfs installation.
2024-02-18T14:06:25.206287839+10:00 Deploying Sysbox installer helper on the host ...
2024-02-18T14:06:25.499708961+10:00 Running Sysbox installer helper on the host (may take several seconds) ...
2024-02-18T14:06:28.052809028+10:00 Stopping the Sysbox installer helper on the host ...
2024-02-18T14:06:28.404940170+10:00 Removing Sysbox installer helper from the host ...
2024-02-18T14:06:28.704056754+10:00 Installing Sysbox on host ...
Configuring host sysctls ...
2024-02-18T14:06:29.921906091+10:00 **sysctl: cannot stat /proc/sys/kernel/unprivileged_userns_clone: No such file or directory**
2024-02-18T14:06:29.922058926+10:00 fs.inotify.max_queued_events = 1048576
2024-02-18T14:06:29.922064366+10:00 fs.inotify.max_user_watches = 1048576
fs.inotify.max_user_instances = 1048576
2024-02-18T14:06:29.922069990+10:00 kernel.keys.maxkeys = 20000
2024-02-18T14:06:29.922072582+10:00 kernel.keys.maxbytes = 1400000
2024-02-18T14:06:29.922075222+10:00 kernel.pid_max = 4194304

### PR DESCRIPTION
**Pull request by zalsader**
Fixes #780 No such file or directory - unprivileged_userns_clone

I'm currently testing this on ubuntu 20.04 with kernel 6.8.10, but I wanted to get this out for feedback and to see if other people wanted to test it.

### CODE REVIEW

File: sysbox-in-docker/sysbox [hunk 1]
```diff
@@ -272,10 +288,28 @@ function sysfs_mount_setup() {
...
+	if [ "$(printf '%s\n' "$requiredver" "$currentver" | sort -V | head -n1)" = "$requiredver" ]; then 
+			# Greater than or equal to $requiredver
+			true
+	else
+			# Less than $requiredver
+			false
+	fi
+}
+
+# Returns linux kernerl version in the system.
```
Comment (dray92) [inline sysbox-in-docker/sysbox:303]:
```suggestion
# Returns linux kernel version in the system.
```

**Comment by ctalledo** [comment]
Hi @zalsader, thanks for the contribution.

Looking at your changes though, I can see that it's not complete. Allow me to explain.

There are 2 main areas that we need to change to adjust for `/proc/sys/kernel/unprivileged_userns_clone` no longer being available in Ubuntu/Debian distros with kernel 6.6+.

* The sysbox package installer, [here](https://github.com/nestybox/sysbox-pkgr/blob/cf5355d2842790515ef26a447e67c8c60b49e975/deb/sysbox-ce/sysbox-ce.postinst#L85) and [here](https://github.com/nestybox/sysbox-pkgr/blob/cf5355d2842790515ef26a447e67c8c60b49e975/systemd/99-sysbox-sysctl.conf#L4).
 
* The sysbox k8s installer, [here](https://github.com/nestybox/sysbox-pkgr/blob/cf5355d2842790515ef26a447e67c8c60b49e975/k8s/systemd/99-sysbox-sysctl.conf#L4).

For all of these, we ideally need to adjust the code to say: "if `/proc/sys/kernel/unprivileged_userns_clone` exists, set it to 1; else, check if `/proc/sys/user/max_user_namespaces` exists and if it's set to 0, set it to 1."

In addition, we can also change the `scr/sysbox` script (the one you touched), though that script is only used in the Sysbox test suite right now (it's not distributed with Sysbox). The change should use the same logic I described above though.

Optionally we can change the sysbox-in-docker/sysbox script too, though that is no longer used anywhere AFAIK.

Let me know if you are up to the task of changing the above files, otherwise we can do it (though can't commit to a date yet).

**Comment by zalsader** [comment]
I changed the code to check if the file exists instead of depending on the kernel version. I also created a PR in `sysbox-pkgr` https://github.com/nestybox/sysbox-pkgr/pull/129 to check if `unprivileged_userns_clone` exists before writing. I could not figure out how to change [this](https://github.com/nestybox/sysbox-pkgr/blob/cf5355d2842790515ef26a447e67c8c60b49e975/k8s/systemd/99-sysbox-sysctl.conf#L4) and [this](https://github.com/nestybox/sysbox-pkgr/blob/cf5355d2842790515ef26a447e67c8c60b49e975/systemd/99-sysbox-sysctl.conf#L4) depending on the existence of the setting, but I think it's ok since the rpm installer uses the same files without modifications, and fedora/centos don't have `kernel.unprivileged_userns_clone`. Currently testing.

**Comment by zalsader** [comment]
Thank you for approving this. It would be nice to be able to include upgrading the `sysbox-pkgr` submodules with https://github.com/nestybox/sysbox-pkgr/pull/129 before merging.

> But question: don't we need to check if `/proc/sys/user/max_user_namespaces` exists and if it's set to 0, set it to 1. IOW, I understood that in newer Ubuntu kernels, `/proc/sys/kernel/unprivileged_userns_clone` was replaced with `/proc/sys/user/max_user_namespaces`.

The [rest of the code](https://github.com/nestybox/sysbox/blob/a6e09bf360edadb528cb71bf39ea5d4367211f24/scr/sysbox#L120-L124) in this repo does that. It could make sense to address this in [sysbox-pkgr](https://github.com/nestybox/sysbox-pkgr), but note that [sysbox-runc](https://github.com/nestybox/sysbox-runc/blob/0ce69d704e991c50011d50b707fde221ad806d63/libsysbox/sysbox/utils.go#L58-L92) errors if userns is not configured. I don't think setting `max_user_namespaces` to 1 is the idiomatic approach, I believe that letting the sysadmin see a configuration error (from runc) then choose the appropriate value would be better.

**Comment by ctalledo** [comment]
> The [rest of the code](https://github.com/nestybox/sysbox/blob/a6e09bf360edadb528cb71bf39ea5d4367211f24/scr/sysbox#L120-L124) in this repo does that. It could make sense to address this in [sysbox-pkgr](https://github.com/nestybox/sysbox-pkgr), but note that `sysbox-runc` [errors if userns is not configured](https://github.com/nestybox/sysbox-runc/blob/0ce69d704e991c50011d50b707fde221ad806d63/libsysbox/sysbox/utils.go#L58-L92). I don't think setting `max_user_namespaces` to 1 is the idiomatic approach, I believe that letting the sysadmin see a configuration error (from runc) then choose the appropriate value would be better.

Yes, makes sense as discussed [here](https://github.com/nestybox/sysbox-pkgr/pull/129#issuecomment-2142697924).

**Pull request merged**